### PR TITLE
Fix prefill

### PR DESCRIPTION
**Pull request by samuki**
`_batch_evaluate` ran one `generate() `over all unique prompts and asserted one captured logprob row per prompt. A large or long batch is scheduled across multiple prefill steps, but `GlobalLogprobsCapture` captures only one step, so the call returns fewer rows than prompts and the assert fails. This crashed spider under SMC.batched.

Fix: split each adapter's prompts into sub-batches that stay within the scheduler's `max_num_batched_tokens` and `max_num_seqs`.

Test: `tests/test_sub_batches.py` covers the packing invariant (max_tokens, max_seqs, oversized prompt, order/coverage).

### CODE REVIEW

**Review by shepardxia — COMMENTED** [review]
`_batch_evaluate` fix is right. Two inline suggestions below extract a shared `_capture_logprobs` so the same fix also covers the **sync** batch path (`batch_next_token_logprobs_sync` / `next_token_logprobs_sync`), which has the identical multi-prefill-step row-drop and isn't touched here.

File: genlm/backend/llm/vllm.py [hunk 1]
```diff
@@ -490,28 +513,31 @@ def _batch_evaluate(self):
...
+                        assert all_logprobs is not None, "Logprobs should be captured"
+                        assert all_logprobs.shape[0] == len(sub), (
+                            f"Expected {len(sub)} logprobs, got {all_logprobs.shape[0]}"
+                        )
+                        for i, key in enumerate(sub):
+                            captured[key] = all_logprobs[i]
+                    for key, logprobs in captured.items():
+                        futures = [f for f in query_groups[key] if not f.done()]
                         if len(futures) == 1:
                             futures[0].set_result(logprobs)
```
Comment (shepardxia) [inline genlm/backend/llm/vllm.py:542]: Extract the generate/capture block into a shared `_capture_logprobs` (defined in the next suggestion); the per-adapter loop is then just resolution. Drops the now-unused `_prefill_step_limits()` call here (the helper computes its own).

```suggestion
            for lora_name, query_groups in by_lora.items():
                unique_token_ids = list(query_groups.keys())
                try:
                    rows = self._capture_logprobs(unique_token_ids, lora_name)
                    for i, key in enumerate(unique_token_ids):
                        logprobs = rows[i]
                        futures = [f for f in query_groups[key] if not f.done()]
                        if len(futures) == 1:
                            futures[0].set_result(logprobs)
```

File: genlm/backend/llm/vllm.py [hunk 2]
```diff
@@ -523,6 +549,21 @@ def _batch_evaluate(self):
...
+            own sub-batch (one prompt always yields a single capture row)."""
+            batch, total = [], 0
+            for t in token_id_tuples:
+                if batch and (total + len(t) > max_tokens or len(batch) >= max_seqs):
+                    yield batch
+                    batch, total = [], 0
+                batch.append(t)
+                total += len(t)
+            if batch:
+                yield batch
```
Comment (shepardxia) [inline genlm/backend/llm/vllm.py:565]: Keep `_sub_batches`, add the shared helper after it. Then route the **sync** paths through it too — they have the identical multi-prefill-step row-drop but aren't in this diff:
- `batch_next_token_logprobs_sync` → `return self._capture_logprobs([tuple(t) for t in token_ids_list], lora_name)`
- `next_token_logprobs_sync` → `self._capture_logprobs([tuple(token_ids)], lora_name)[0]`

```suggestion
        @staticmethod
        def _sub_batches(token_id_tuples, max_tokens, max_seqs):
            """Greedily pack prompts into sub-batches of ``<= max_tokens`` total tokens
            AND ``<= max_seqs`` prompts. A prompt longer than ``max_tokens`` becomes its
            own sub-batch (one prompt always yields a single capture row)."""
            batch, total = [], 0
            for t in token_id_tuples:
                if batch and (total + len(t) > max_tokens or len(batch) >= max_seqs):
                    yield batch
                    batch, total = [], 0
                batch.append(t)
                total += len(t)
            if batch:
                yield batch

        def _capture_logprobs(self, token_id_tuples, lora_name):
            """[len(prompts), V] logprobs in input order, under one adapter. Sub-batches
            so each generate() is one prefill step -> one capture row per prompt."""
            if not token_id_tuples:
                raise ValueError("no prompts to evaluate")
            max_tokens, max_seqs = self._prefill_step_limits()
            rows = []
            for sub in self._sub_batches(token_id_tuples, max_tokens, max_seqs):
                self.logprobs_capture.clear()
                self.llm_engine.generate(
                    prompts=[TokensPrompt(prompt_token_ids=list(t)) for t in sub],
                    sampling_params=SamplingParams(**self.default_params),
                    lora_request=self._lora_request_for(lora_name),
                    use_tqdm=False,
                )
                all_logprobs = self.logprobs_capture.get_all_logprobs()
                assert all_logprobs is not None, "Logprobs should be captured"
                assert all_logprobs.shape[0] == len(sub), (
                    f"Expected {len(sub)} logprobs, got {all_logprobs.shape[0]}"
                )
                rows.append(all_logprobs)
            return rows[0] if len(rows) == 1 else torch.cat(rows, dim=0)
```

**Comment by samuki** [comment]
Moved to the PR on the main branch #77
